candle-core/src/metal_backend/device.rs

```diff
@@ -26,6 +29,106 @@ impl DeviceId {
     }
 }
 
+#[derive(Clone)]
+pub(crate) struct AllocationPolicy {
+    /// Total bytes we can allocate before forcing a sync to reclaim temporaries.
+    pending_limit_bytes: usize,
+    /// Maximum bytes to keep cached for reuse.
+    cache_limit_bytes: usize,
+}
+
+impl Default for AllocationPolicy {
+    fn default() -> Self {
+        const DEFAULT_PENDING: usize = 4 * 1024 * 1024 * 1024; // 4 GiB
+        const MIN_PENDING: usize = 512 * 1024 * 1024; // 512 MiB
+        const MAX_PENDING: usize = 12 * 1024 * 1024 * 1024; // 12 GiB
+
+        fn parse_env_mebibytes(var: &str) -> Option<usize> {
+            std::env::var(var)
+                .ok()
+                .and_then(|value| value.trim().parse::<usize>().ok())
+                .map(|mb| mb * 1024 * 1024)
+        }
+        fn sysctl_u64(name: &[u8]) -> Option<u64> {
+            use libc::c_void;
+            unsafe {
+                let mut len: usize = 0;
+                if libc::sysctlbyname(
+                    name.as_ptr() as *const libc::c_char,
+                    std::ptr::null_mut(),
+                    &mut len as *mut usize,
+                    std::ptr::null_mut(),
+                    0,
+                ) != 0
+                {
+                    return None;
+                }
+                if len == 0 || len > core::mem::size_of::<u64>() {
+                    return None;
+                }
+                let mut value: u64 = 0;
+                if libc::sysctlbyname(
+                    name.as_ptr() as *const libc::c_char,
+                    &mut value as *mut u64 as *mut c_void,
+                    &mut len as *mut usize,
+                    std::ptr::null_mut(),
+                    0,
+                ) != 0
+                {
+                    return None;
+                }
+                Some(value)
+            }
+        }
+
+        fn system_memory_bytes() -> Option<usize> {
+            const MEBIBYTE: usize = 1024 * 1024;
+            const SYSTEM_RESERVE_FRACTION: usize = 4; // Keep at least 25% for the OS.
+            const SYSTEM_RESERVE_MIN: usize = 2 * 1024 * 1024 * 1024; // 2 GiB floor.
+
+            let hw_total = sysctl_u64(b"hw.memsize\0").and_then(|bytes| {
+                if bytes == 0 || bytes > usize::MAX as u64 {
+                    None
+                } else {
+                    Some(bytes as usize)
+                }
+            })?;
```

Comment:
We have to go all the way back to 2008 to worry about wether our apple products are 32bit or 64bit.

```suggestion
            let hw_total = sysctl_u64(HW_MEMSIZE_KEY)? as usize;
```